File: ios/HolyDay/Services/Common/WhatsNewService.swift
//
//  WhatsNewService.swift
//  HolyDay
//

import Foundation
import Observation

/// Décide si l'écran de nouveautés doit s'afficher après une mise à jour, et pour quelles versions.
///
/// Deux invariants tiennent tout le comportement :
///
/// 1. **Une installation neuve ne voit jamais de nouveautés.** L'onboarding appelle `markSeen()` à
///    sa dernière étape : la version courante est donc déjà enregistrée quand l'utilisateur entre
///    dans l'app. Sans ça, un nouvel utilisateur se verrait annoncer comme « nouveau » ce qu'il
///    découvre de toute façon pour la première fois.
/// 2. **Aucune version marquée signifie « mise à jour depuis un binaire antérieur à cette
///    fonctionnalité »**, et non « installation neuve » — d'où l'affichage dans ce cas. Le point 1
///    garantit que les deux situations ne se confondent pas.
///
/// Les versions sautées sont rattrapées : quelqu'un qui passe de 1.0.1 à 1.2 voit les notes de 1.1
/// et de 1.2 dans le même écran.
///
/// Entrées injectables (stockage, version courante, catalogue) pour rendre la décision testable
/// sans toucher au bundle ni aux préférences réelles.
@MainActor
@Observable
final class WhatsNewService {
  static let shared = WhatsNewService()

  /// Non-nil quand un écran de nouveautés est en attente d'affichage.
  private(set) var pending: Presentation?

  /// Regroupe les notes à présenter en une valeur `Identifiable`, pour `sheet(item:)` — qui, au
  /// contraire de `sheet(isPresented:)`, conserve son contenu pendant l'animation de fermeture.
  struct Presentation: Identifiable {
    let releases: [ReleaseNote]
    var id: String { releases.map(\.version).joined(separator: "+") }
  }

  private let defaults: UserDefaults
  private let currentVersion: String
  private let catalog: [ReleaseNote]

  private let lastSeenKey = "holyday.whatsNew.lastSeenVersion"

  init(
    defaults: UserDefaults = .standard,
    currentVersion: String = Bundle.main.appVersion,
    catalog: [ReleaseNote] = ReleaseNotesCatalog.all
  ) {
    self.defaults = defaults
    self.currentVersion = currentVersion
    self.catalog = catalog
  }

  // MARK: - Décision

  /// À appeler une fois le splash retiré, pour ne pas présenter une feuille sous l'écran de
  /// lancement. Sans effet en mode capture d'écran : les runs `fastlane snapshot` doivent rester
  /// déterministes.
  func evaluate() {
    guard !ScreenshotMode.isActive else { return }

    let lastSeen = defaults.string(forKey: lastSeenKey)
    let releases = releasesNewer(than: lastSeen)

    guard !releases.isEmpty else {
      // Rien à annoncer pour cette version : on avance quand même le repère, sinon la note d'une
      // version ultérieure ressortirait accompagnée de celles, périmées, des versions traversées.
      markSeen()
      return
    }

    pending = Presentation(releases: releases)
  }

  /// Notes strictement postérieures à `lastSeen`, plus récente en premier.
  /// `lastSeen == nil` (mise à jour depuis un binaire antérieur à cette fonctionnalité) ne renvoie
  /// que la note de la version installée : annoncer tout l'historique n'aurait aucun sens.
  private func releasesNewer(than lastSeen: String?) -> [ReleaseNote] {
    guard let lastSeen else {
      return catalog.filter { $0.version == currentVersion }
    }
    return
      catalog
      .filter { note in
        Self.isVersion(note.version, newerThan: lastSeen)
          && !Self.isVersion(note.version, newerThan: currentVersion)
      }
      .sorted { Self.isVersion($0.version, newerThan: $1.version) }
  }

  /// Comparaison numérique composant par composant : « 1.10 » est postérieure à « 1.9 », ce qu'une
  /// comparaison lexicographique donnerait à l'envers.
  static func isVersion(_ lhs: String, newerThan rhs: String) -> Bool {
    lhs.compare(rhs, options: .numeric) == .orderedDescending
  }

  // MARK: - Mutations

  /// Enregistre la version installée comme vue et referme l'écran.
  /// Appelé à la fermeture de la feuille **et** à la fin de l'onboarding (cf. invariant 1).
  func markSeen() {
    defaults.set(currentVersion, forKey: lastSeenKey)
    pending = nil
  }

  /// Efface le repère : au prochain `evaluate()`, les nouveautés de la version installée
  /// ressortent. Réservé au menu développeur — délibérément **pas** appelé par l'effacement
  /// complet des données, qui porte sur les données de l'utilisateur, pas sur le binaire installé :
  /// réafficher les nouveautés d'une version déjà utilisée depuis des semaines n'aurait aucun sens.
  func reset() {
    defaults.removeObject(forKey: lastSeenKey)
    pending = nil
  }
}

extension Bundle {
  /// `CFBundleShortVersionString` (« 1.0.1 »), repli sur « 0 » — une version illisible ne doit pas
  /// faire croire à une mise à jour à chaque lancement.
  var appVersion: String {
    object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0"
  }
}

File: android/app/src/main/java/com/matthiascadet/holyday/service/WhatsNewService.kt
package com.matthiascadet.holyday.service

import android.content.SharedPreferences
import com.matthiascadet.holyday.data.model.ReleaseNote
import com.matthiascadet.holyday.data.model.ReleaseNotesCatalog
import com.matthiascadet.holyday.data.prefs.AppPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Décide quelles notes de version annoncer après une mise à jour.
 *
 * L'onboarding appelle [markSeen] avant d'ouvrir l'écran principal : une installation neuve ne
 * voit donc jamais des nouveautés qu'elle découvre pour la première fois. L'absence de repère
 * correspond ainsi sans ambiguïté à une mise à jour depuis un binaire plus ancien.
 */
class WhatsNewService(
    private val preferences: SharedPreferences = AppPreferences.raw,
    private val currentVersion: () -> String,
    private val catalog: List<ReleaseNote> = ReleaseNotesCatalog.all,
) {
    private val _pending = MutableStateFlow<List<ReleaseNote>?>(null)
    val pending: StateFlow<List<ReleaseNote>?> = _pending.asStateFlow()

    fun evaluate() {
        val version = currentVersion()
        val lastSeen = preferences.getString(LAST_SEEN_KEY, null)
        val releases = releasesNewerThan(lastSeen, version)
        if (releases.isEmpty()) {
            markSeen()
        } else {
            _pending.value = releases
        }
    }

    fun markSeen() {
        preferences.edit().putString(LAST_SEEN_KEY, currentVersion()).apply()
        _pending.value = null
    }

    /** Réservé au menu développeur, comme sur iOS. */
    fun reset() {
        preferences.edit().remove(LAST_SEEN_KEY).apply()
        _pending.value = null
    }

    private fun releasesNewerThan(lastSeen: String?, version: String): List<ReleaseNote> =
        if (lastSeen == null) {
            catalog.filter { it.version == version }
        } else {
            catalog
                .filter { isVersionNewer(it.version, lastSeen) && !isVersionNewer(it.version, version) }
                .sortedWith { first, second -> if (isVersionNewer(first.version, second.version)) -1 else 1 }
        }

    companion object {
        const val LAST_SEEN_KEY = "holyday.whatsNew.lastSeenVersion"

        val shared: WhatsNewService by lazy {
            WhatsNewService(currentVersion = { com.matthiascadet.holyday.BuildConfig.VERSION_NAME.substringBefore('-') })
        }

        /** Comparaison numérique : 1.10 est bien postérieure à 1.9. */
        fun isVersionNewer(lhs: String, rhs: String): Boolean {
            val left = lhs.split('.').map { it.toIntOrNull() ?: 0 }
            val right = rhs.split('.').map { it.toIntOrNull() ?: 0 }
            val size = maxOf(left.size, right.size)
            return (0 until size).firstOrNull { index ->
                (left.getOrElse(index) { 0 }) != (right.getOrElse(index) { 0 })
            }?.let { index -> left.getOrElse(index) { 0 } > right.getOrElse(index) { 0 } } ?: false
        }
    }
}
